applications/asset_tracker_v2/src/modules/util_module.c

```diff
@@ -199,127 +203,133 @@ static void send_reboot_request(void)
 
 		EVENT_SUBMIT(util_module_event);
 
+		state_set(STATE_REBOOT_PENDING);
+
 		error_signaled = true;
 	}
 }
 
-/* Message handler for all states. */
-static void on_all_states(struct util_msg_data *msg)
+/* This API should be called exactly once for each _SHUTDOWN_READY event received from active
+ * modules in the application. When this API has been called a set number of times equal to the
+ * number of active modules, a reboot will be schedueled.
+ */
+static void reboot_ack_check(void)
 {
-	static int reboot_ack_cnt;
-
-	if (is_cloud_module_event(&msg->module.cloud.header)) {
-		switch (msg->module.cloud.type) {
-		case CLOUD_EVT_ERROR:
-			/* Fall through. */
-		case CLOUD_EVT_FOTA_DONE:
-			send_reboot_request();
-			break;
-		case CLOUD_EVT_SHUTDOWN_READY:
-			reboot_ack_cnt++;
-			break;
-		default:
-			break;
-		}
+	static uint8_t reboot_ack_count;
+
+	reboot_ack_count++;
+
+	/* Reboot after a shorter timeout if all modules have acknowledged that they are ready
+	 * to reboot, ensuring a graceful shutdown.
+	 */
+	if (reboot_ack_count >= module_active_count_get() - 1) {
+		LOG_WRN("All modules have ACKed the reboot request.");
+		LOG_WRN("Reboot in 5 seconds.");
+		k_delayed_work_submit(&reboot_work, K_SECONDS(5));
 	}
+}
 
-	if (is_modem_module_event(&msg->module.modem.header)) {
-		switch (msg->module.modem.type) {
-		case MODEM_EVT_ERROR:
-			send_reboot_request();
-			break;
-		case MODEM_EVT_SHUTDOWN_READY:
-			reboot_ack_cnt++;
-			break;
-		default:
-			break;
-		}
+/* Message handler for STATE_INIT. */
+static void on_state_init(struct util_msg_data *msg)
+{
+	if ((IS_EVENT(msg, cloud, CLOUD_EVT_ERROR)) ||
```

Comment:
I know that it's not a functionality change, but Is really every `CLOUD_EVT_ERROR` and `MODEM_EVT_ERROR` unrecoverable and can only be handled by a reboot?

Reply:
Yes, all errors sent to the utility module are considered unrecoverable.